mcs/class/corlib/Test/System.Reflection.Emit/ModuleBuilderTest.cs

```diff
@@ -806,5 +805,64 @@ public void GetType ()
 			Assert.AreEqual ("t1&", module.GetType ("t1&").FullName);
 			Assert.AreEqual ("t1[]&", module.GetType ("t1[]&").FullName);
 		}
+
+		[AttributeUsage(AttributeTargets.All)]
+		public class MyAttribute : Attribute {
+			String s;
+			public MyAttribute (String s) 
+			{
+				this.s = s;
+			}
+		}
+
+		[Test]
+		public void GetMethodsBeforeInstantiation ()
+		{
+			AssemblyBuilder assm = AssemblyBuilder.DefineDynamicAssembly (new AssemblyName ("Name"), AssemblyBuilderAccess.Run);
+			ModuleBuilder module = assm.DefineDynamicModule ("Module");
+			FieldBuilder fieldBuilder = module.DefineInitializedData ("GlobalField", new byte[4], FieldAttributes.Public);
+
+			MethodBuilder method = module.DefinePInvokeMethod ("printf", "libc.so",
+															  MethodAttributes.PinvokeImpl | MethodAttributes.Static | MethodAttributes.Public,
+															  CallingConventions.Standard, typeof (void), new Type [] { typeof (string) }, CallingConvention.Winapi,
+															  CharSet.Auto);
+			method.SetImplementationFlags (MethodImplAttributes.PreserveSig |
+										   method.GetMethodImplementationFlags ());
+
+			module.CreateGlobalFunctions ();
+
+			Assert.AreEqual (module.GetMethods (BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance).Length, 1);
+		}
+
+		[Test]
+		public void GetFieldsBeforeInstantiation ()
+		{
+			AssemblyBuilder assm = AssemblyBuilder.DefineDynamicAssembly (new AssemblyName ("Name"), AssemblyBuilderAccess.Run);
+			ModuleBuilder module = assm.DefineDynamicModule ("Module");
+			FieldBuilder fieldBuilder = module.DefineInitializedData ("GlobalField", new byte[4], FieldAttributes.Public);
+			module.CreateGlobalFunctions ();
+
+			var fieldG = module.GetField (fieldBuilder.Name);
+			Assert.IsNotNull (fieldG);
+			Assert.AreEqual (module.GetFields (BindingFlags.Public | BindingFlags.Static | BindingFlags.Instance).Length, 1);
+		}
+
+		[Test]
+		public void GetCustomAttributesBeforeInstantiation ()
+		{
+			AssemblyBuilder assm = AssemblyBuilder.DefineDynamicAssembly (new AssemblyName ("Name"), AssemblyBuilderAccess.Run);
+			ModuleBuilder module = assm.DefineDynamicModule ("Module");
+			FieldBuilder fieldBuilder = module.DefineInitializedData ("GlobalField", new byte[4], FieldAttributes.Public);
+			module.CreateGlobalFunctions ();
+
+			ConstructorInfo ctor = typeof(MyAttribute).GetConstructor (new Type [] {typeof(String)});
+			ctor.GetHashCode ();
+			CustomAttributeBuilder cab = new CustomAttributeBuilder (ctor, new object [] {"hi"});
+			module.SetCustomAttribute (cab);
+
+			Assert.AreEqual (module.GetCustomAttributes (false).Length, 1);
+			Assert.AreEqual (module.GetCustomAttributes (false)[0].GetType (), typeof (MyAttribute));
+		}
+
```

Comment:
It would be good to also add a test where a `TypeBuilder` is created in `module` and then a `ConstructorInfo` from that `TypeBuilder` is used for a `CustomAttributeBuilder` that you then hang on the module.

Reply:
Also possibly something with two separate `ModuleBuilder`s where a `ConstructorInfo` from `module1` is used for an attribute on `module2`.  (In this case I'd want to check the `Run` behavior is ok, and also a test for `Save()` followed by an assembly load to make sure we wrote reasonable typeref tokens)